### ISSUE TITLE
Add snapshot methods to query builder/model

### ISSUE THREAD

**Issue by matthewjumpsoffbuildings** [colopl/laravel-spanner#231]
Would be great if the snapshot method could be chained directly on query builder or model method calls. I tried to suss it myself to make a PR but I am not sure the best approach.

Ideally it would be great if you could do something like

```php
// snapshot from a model
User::snapshot()->all();

// snapshot from a builder
$queryBuilder
    ->where('foo', 'bar')
    ->snapshot()
    ->get();
```

Ideally you should be able to insert the `snapshot()` method anywhere in the chain, but since snapshots use a closure to execute under the hood Im not sure how it could be made chainable

**Comment by taka-oyama** [comment]
A normal SELECT statement outside of a transaction is already a read only transaction.

**Comment by matthewjumpsoffbuildings** [comment]
So running a SELECT sets `currentSnapshot` somewhere on the Connection and makes it execute the query in a snapshot? I am struggling to see where this happens in the code?

I remember there was some confusion between snapshots and timestamp bounds/single use transactions. Timestamp bounds and single use transactions are not the same as snapshots, as discussed here:

- https://github.com/colopl/laravel-spanner/issues/145#issuecomment-2143387161
- https://github.com/colopl/laravel-spanner/issues/145#issuecomment-2144377976
- https://github.com/colopl/laravel-spanner/issues/145#issuecomment-2144379350

**Comment by matthewjumpsoffbuildings** [comment]
Also if it is the case that normal SELECT statements do somehow set `currentSnapshot`, how do you specify the timestamp bounds for the snapshot? Because the only place I see a snapshot being created is in the `ManagesSnapshots` trait, and the only time that happens is when `Connection::snapshot($timestampBound)` is called directly, which is surely not called when you just run a normal SELECT?

**Comment by matthewjumpsoffbuildings** [comment]
I just tested this, if I run something like

```php
DB::connection()->table('totals')->get();
```

At no point does `Connection::executeSnapshotQuery` get called, nor does `Connection::snapshot`. Since `Connection::snapshot` seems to be the only place that actually runs `this->getSpannerDatabase()->snapshot($options);`, it is not the case that normal SELECT statements use snapshots. Rather, as before, they appear to be using single use transactions (optionally with timestamp bounds), which as discussed previously, are not the same as snapshots, and do not _guarantee_ lock free reads

**Comment by matthewjumpsoffbuildings** [comment]
Finally, it should actually not be the case that regular SELECT statements use snapshots. Snapshot use should require explicit specification.

This code _should not_ run in a snapshot:
```php
DB::connection()->table('totals')->get();
```

However something like this should be possible
```php
DB::connection()->table('totals')->snapshot(new ExactStaleness(20))->get();
```

**Comment by taka-oyama** [comment]
From what I can read from [here](https://cloud.google.com/spanner/docs/reference/rest/v1/TransactionOptions#:~:text=Cloud%20Spanner%20supports%20three%20transaction%20modes%3A), all non read write transactions are snapshot read-only transactions.

So running [this](https://github.com/colopl/laravel-spanner/blob/master/src/Connection.php#L592) is automatically running the statement as singleUse snapshot read-only transaction. The only reason I added `Connection::snapshot(...)` to connection is to make it possible to set the staleness for all queries inside the closure for convenience.

**Comment by matthewjumpsoffbuildings** [comment]
I am confused by this. It appears perhaps the documentation is contradicting itself when it says [this](https://cloud.google.com/spanner/docs/timestamp-bounds#timestamp_bound_types:~:text=Although%20reads%20using%20these%20timestamp%20bound%20modes%20are%20not%20part%20of%20a%20read%2Dwrite%20transaction%2C%20they%20can%20block%20waiting%20for%20concurrent%20read%2Dwrite%20transactions%20to%20commit.%20Bounded%20staleness%20reads%20attempt%20to%20pick%20a%20timestamp%20to%20avoid%20blocking%2C%20but%20may%20still%20block.)

> Although reads using these timestamp bound modes are not part of a read-write transaction, they can block waiting for concurrent read-write transactions to commit. Bounded staleness reads attempt to pick a timestamp to avoid blocking, but may still block.

So its saying you can have a stale read that is not part of a read-write transaction (which according to the documentation you linked saying theres only 3 transaction modes, read-write, snapshot, partitioned, must mean that its in a snapshot right, since it cant be partitioned), and yet this read that is in a snapshot can still lock?

When the _other_ documentation says snapshot _never_ lock, as stated [here](https://cloud.google.com/spanner/docs/reference/rest/v1/TransactionOptions#:~:text=Snapshot%20transactions%20do%20not%20take%20locks.%20Instead%2C%20they%20work%20by%20choosing%20a%20Cloud%20Spanner%20timestamp%2C%20then%20executing%20all%20reads%20at%20that%20timestamp.%20Since%20they%20do%20not%20acquire%20locks%2C%20they%20do%20not%20block%20concurrent%20read%2Dwrite%20transactions)

> Snapshot transactions do not take locks. Instead, they work by choosing a Cloud Spanner timestamp, then executing all reads at that timestamp. Since they do not acquire locks, they do not block concurrent read-write transactions.

From my understanding, unless you explicitly call `Database::snapshot()` to create a `Snapshot` object, a read you run could still potentially lock, even with timestamp bounds

Perhaps this is a case of the documentation being unclear

**Comment by taka-oyama** [comment]
> Bounded staleness reads attempt to pick a timestamp to avoid blocking, but may still block.

The doc really should go more into detail about how it can still block.
From what I have tested in production, switching strong reads to stale reads resulted in better performance and less blocking overall.

**Comment by matthewjumpsoffbuildings** [comment]
> The doc really should go more into detail about how it can still block.

Agreed, it seems confusing and like a direct contradiction currently, but it really does seem to me that the best practice if you _want_ snapshot reads with a complete guarantee of no locking, that you should call `Database::snapshot()` and run `execute()` on the returned `Snapshot` object

> From what I have tested in production, switching from strong reads to stale reads resulted in better performance and less blocking overall.

I am sure it does, but since the docs do seem to indicate that stale reads alone is not enough to guarantee no locks, it would make sense to add explicit snapshot usage options

Your closure method is enough for me to have peace of mind, but adding the ability to trigger implicit snapshots on `Builder` and `Model` would be a nice to have
